design/drive-status.md

```diff
@@ -0,0 +1,141 @@
+
+## Motivation
+
+We need to be able to collect and update enough information about hardware
+devices in a server to monitor their status and identify them for
+maintenance and replacement.
+Valuable information are for example the current status, type, serial number,
+physical location, part number.
+We need to be able to refresh the information about the status of the hardware
+devices and, if a device has been replaced, we need to be able to retrieve the
+new information and update the hardware inventory to provide a constant
+up-to-date vision on the status of the devices.
+
+### Goals
+
+1. Decide which information on the devices to retrieve.
+2. Agree on how to retrieve such information.
+3. Provide a way to manually/automatically collect and update the devices
+information.
+
+### Non-Goals
+
+1. Any update on device data should not trigger any reaction from ironic.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+- Depending on the device, we need to collect the following information, when
+applicable:
+    * Capacity
+    * Location
+    * Manufacturer
+    * Model
+    * Part Number
+    * Serial Number
+    * Status
+    * SMART details
+    
+- NVMe drives require special tools to get SMART data (e.g. nvme-cli),
+although [smartmontools](https://www.smartmontools.org/wiki/NVMe_Support)
+supports NVMe devices since version 6.5.
+ 
+- [Hdparm](https://en.wikipedia.org/wiki/Hdparm) can provide SMART data for
+most of drives. It doesn't work with NVMe devices.
+
+- [python-hardware](https://github.com/redhat-cip/hardware) can collect SMART
+data, although it requires code changes to detect and provide info on NVMe
+devices, for example relying directly on nvme tools instead of smartmontools.
+
+- [Ironic Python Agent (IPA)](https://github.com/openstack/ironic-python-agent)
+is already able to collect most of this information.
+Using a subset of the IPA code running inside a container in the hosts where we
+want to collect the drives data from can also allow to provide periodic
+updates on all the required info.
+The new containerized service will talk to Ironic Inspector.
+
+- [Ironic Inspector](https://docs.openstack.org/ironic-inspector/latest/)
+needs to be able to publish up-to-date data at any time, including from active
+and available nodes.
+This feature requires changes in the inspector code that currently accept data
+updates only from nodes in manageable state.
+
+### Risks and Mitigations
+
+1. Upstream Ironic community may reject changes in Ironic Inspector code.
+
+## Design Details
+
+### Work Items
+
+- Add option to detect and provide info on NVMe drives to python-hardware.
+- Extract and modify part of the IPA code to be able to run it in a container
+as a new service.
+- Modify Ironic Inspector code to accept data from active nodes.
+- Package (containerize) the new service based on IPA and the necessary tools,
+specifically for NVMe compatibility.
+- Distribute the new service to the hosts. 
```

Comment:
I think we need to add more detail around this, because it's going to involve making some decisions about which component is responsible for ensuring that it happens.

I could see the baremetal operator setting up the DaemonSet to ensure the agent is deployed. However, the operator does not currently assume that it is deploying members of any cluster, or that members of the current cluster were deployed on baremetal.

The baremetal provider for the machine-api could set up the DaemonSet, but it does not currently know anything about the fact that we are using Ironic so it would be a little odd for it to do so.

It seems like we need something that knows it is managing a baremetal cluster using those other 2 components, and that it should set up monitoring on that cluster by creating the DaemonSet for the agent as one of the components.

Reply:
I think this was already addressed in a more recent commit.
Please re-open it if it's not the case.